File: server/temporal-keywords.ts
/**
 * Temporal Keyword Expansion Module
 * 
 * Generates hypotheses based on trending keywords during Bitcoin's early years.
 * Words and phrases that were culturally relevant during wallet creation
 * are more likely to have been used as passphrases.
 * 
 * Organized by year and category:
 * - Political events
 * - Technology trends
 * - Pop culture
 * - Economic events
 * - Social phenomena
 */

export interface TemporalKeyword {
  keyword: string;
  year: number;
  category: 'politics' | 'technology' | 'pop_culture' | 'economics' | 'social' | 'crypto';
  relevance: number; // 0-1, how likely to be used as passphrase
  context: string; // Why this was significant
}

/**
 * 2009 Temporal Keywords
 */
export const KEYWORDS_2009: TemporalKeyword[] = [
  // Politics
  { keyword: 'obama', year: 2009, category: 'politics', relevance: 0.7, context: 'Obama inauguration January 2009' },
  { keyword: 'inauguration', year: 2009, category: 'politics', relevance: 0.5, context: 'Historic inauguration' },
  { keyword: 'change we can believe in', year: 2009, category: 'politics', relevance: 0.6, context: 'Obama campaign slogan' },
  { keyword: 'yes we can', year: 2009, category: 'politics', relevance: 0.6, context: 'Obama slogan' },
  
  // Economics
  { keyword: 'bailout', year: 2009, category: 'economics', relevance: 0.8, context: 'Bank bailouts, inspired Bitcoin' },
  { keyword: 'financial crisis', year: 2009, category: 'economics', relevance: 0.7, context: '2008-2009 crisis' },
  { keyword: 'recession', year: 2009, category: 'economics', relevance: 0.6, context: 'Great Recession' },
  { keyword: 'too big to fail', year: 2009, category: 'economics', relevance: 0.7, context: 'Bank phrase' },
  { keyword: 'lehman brothers', year: 2009, category: 'economics', relevance: 0.6, context: 'Bankruptcy 2008' },
  
  // Technology
  { keyword: 'iphone 3gs', year: 2009, category: 'technology', relevance: 0.5, context: 'Released June 2009' },
  { keyword: 'windows 7', year: 2009, category: 'technology', relevance: 0.5, context: 'Released October 2009' },
  { keyword: 'twitter', year: 2009, category: 'technology', relevance: 0.5, context: 'Growing rapidly' },
  { keyword: 'netbook', year: 2009, category: 'technology', relevance: 0.4, context: 'Popular device type' },
  
  // Pop Culture
  { keyword: 'avatar', year: 2009, category: 'pop_culture', relevance: 0.6, context: 'Movie released December 2009' },
  { keyword: 'michael jackson', year: 2009, category: 'pop_culture', relevance: 0.6, context: 'Died June 2009' },
  { keyword: 'swine flu', year: 2009, category: 'social', relevance: 0.5, context: 'H1N1 pandemic' },
  { keyword: 'miracle on the hudson', year: 2009, category: 'social', relevance: 0.5, context: 'January 2009 plane landing' },
  
  // Crypto
  { keyword: 'satoshi', year: 2009, category: 'crypto', relevance: 0.9, context: 'Bitcoin creator' },
  { keyword: 'genesis block', year: 2009, category: 'crypto', relevance: 0.8, context: 'First Bitcoin block' },
  { keyword: 'chancellor brink', year: 2009, category: 'crypto', relevance: 0.8, context: 'Genesis block headline' },
];

/**
 * 2010 Temporal Keywords
 */
export const KEYWORDS_2010: TemporalKeyword[] = [
  // Technology
  { keyword: 'ipad', year: 2010, category: 'technology', relevance: 0.6, context: 'Released April 2010' },
  { keyword: 'instagram', year: 2010, category: 'technology', relevance: 0.5, context: 'Launched October 2010' },
  { keyword: 'foursquare', year: 2010, category: 'technology', relevance: 0.4, context: 'Popular check-in app' },
  
  // Crypto
  { keyword: 'pizza day', year: 2010, category: 'crypto', relevance: 0.8, context: '10,000 BTC for pizza May 22' },
  { keyword: 'laszlo', year: 2010, category: 'crypto', relevance: 0.6, context: 'Pizza buyer' },
  { keyword: 'bitcoin faucet', year: 2010, category: 'crypto', relevance: 0.7, context: 'Gavin Andresen faucet' },
  { keyword: 'slush pool', year: 2010, category: 'crypto', relevance: 0.6, context: 'First mining pool' },
  { keyword: 'wikileaks', year: 2010, category: 'crypto', relevance: 0.7, context: 'Bitcoin donations controversy' },
  
  // Pop Culture
  { keyword: 'inception', year: 2010, category: 'pop_culture', relevance: 0.5, context: 'Popular movie' },
  { keyword: 'fifa world cup', year: 2010, category: 'pop_culture', relevance: 0.4, context: 'South Africa 2010' },
  { keyword: 'vuvuzela', year: 2010, category: 'pop_culture', relevance: 0.4, context: 'World Cup horn' },
  
  // Social
  { keyword: 'deepwater horizon', year: 2010, category: 'social', relevance: 0.5, context: 'Oil spill disaster' },
  { keyword: 'haiti earthquake', year: 2010, category: 'social', relevance: 0.4, context: 'January 2010 disaster' },
];

/**
 * 2011 Temporal Keywords
 */
export const KEYWORDS_2011: TemporalKeyword[] = [
  // Crypto
  { keyword: 'silk road', year: 2011, category: 'crypto', relevance: 0.7, context: 'Darknet market launched' },
  { keyword: 'one dollar', year: 2011, category: 'crypto', relevance: 0.7, context: 'BTC reached $1' },
  { keyword: 'parity', year: 2011, category: 'crypto', relevance: 0.6, context: 'BTC = USD parity' },
  { keyword: 'mtgox', year: 2011, category: 'crypto', relevance: 0.8, context: 'Major exchange' },
  
  // Social
  { keyword: 'arab spring', year: 2011, category: 'social', relevance: 0.6, context: 'Middle East uprisings' },
  { keyword: 'occupy wall street', year: 2011, category: 'social', relevance: 0.7, context: 'Financial protest movement' },
  { keyword: 'we are the 99', year: 2011, category: 'social', relevance: 0.6, context: 'Occupy slogan' },
  { keyword: 'fukushima', year: 2011, category: 'social', relevance: 0.5, context: 'Nuclear disaster' },
  { keyword: 'bin laden', year: 2011, category: 'politics', relevance: 0.5, context: 'Death May 2011' },
  
  // Technology
  { keyword: 'iphone 4s', year: 2011, category: 'technology', relevance: 0.5, context: 'Released October 2011' },
  { keyword: 'siri', year: 2011, category: 'technology', relevance: 0.5, context: 'Voice assistant debut' },
  { keyword: 'google plus', year: 2011, category: 'technology', relevance: 0.4, context: 'Launched June 2011' },
  
  // Pop Culture
  { keyword: 'game of thrones', year: 2011, category: 'pop_culture', relevance: 0.5, context: 'TV series premiered' },
  { keyword: 'steve jobs', year: 2011, category: 'pop_culture', relevance: 0.6, context: 'Died October 2011' },
];

/**
 * 2012 Temporal Keywords
 */
export const KEYWORDS_2012: TemporalKeyword[] = [
  // Crypto
  { keyword: 'first halving', year: 2012, category: 'crypto', relevance: 0.8, context: 'November 2012' },
  { keyword: 'halving day', year: 2012, category: 'crypto', relevance: 0.7, context: 'Block reward cut to 25' },
  { keyword: 'asic mining', year: 2012, category: 'crypto', relevance: 0.7, context: 'ASIC miners announced' },
  { keyword: 'butterfly labs', year: 2012, category: 'crypto', relevance: 0.6, context: 'ASIC manufacturer' },
  { keyword: 'coinbase', year: 2012, category: 'crypto', relevance: 0.7, context: 'Exchange founded' },
  { keyword: 'bitcoin foundation', year: 2012, category: 'crypto', relevance: 0.6, context: 'Founded September' },
  
  // Social/Politics
  { keyword: 'mayan calendar', year: 2012, category: 'social', relevance: 0.5, context: '2012 apocalypse myth' },
  { keyword: 'gangnam style', year: 2012, category: 'pop_culture', relevance: 0.6, context: 'Viral sensation' },
  { keyword: 'hurricane sandy', year: 2012, category: 'social', relevance: 0.4, context: 'October 2012 storm' },
  { keyword: 'london olympics', year: 2012, category: 'pop_culture', relevance: 0.4, context: 'Summer 2012' },
];

/**
 * 2013 Temporal Keywords
 */
export const KEYWORDS_2013: TemporalKeyword[] = [
  // Crypto
  { keyword: 'one thousand', year: 2013, category: 'crypto', relevance: 0.8, context: 'BTC hit $1000' },
  { keyword: 'cyprus crisis', year: 2013, category: 'crypto', relevance: 0.7, context: 'Bank crisis, BTC interest spike' },
  { keyword: 'silk road raid', year: 2013, category: 'crypto', relevance: 0.6, context: 'FBI shutdown October' },
  { keyword: 'ross ulbricht', year: 2013, category: 'crypto', relevance: 0.5, context: 'Silk Road arrest' },
  { keyword: 'bip32', year: 2013, category: 'crypto', relevance: 0.6, context: 'HD wallet standard' },
  { keyword: 'bip39', year: 2013, category: 'crypto', relevance: 0.6, context: 'Mnemonic standard' },
  
  // Pop Culture
  { keyword: 'snowden', year: 2013, category: 'politics', relevance: 0.7, context: 'NSA whistleblower' },
  { keyword: 'prism', year: 2013, category: 'politics', relevance: 0.6, context: 'NSA surveillance program' },
  { keyword: 'breaking bad', year: 2013, category: 'pop_culture', relevance: 0.5, context: 'Final season' },
  { keyword: 'bitcoin accepted here', year: 2013, category: 'crypto', relevance: 0.6, context: 'Merchant adoption growing' },
];

/**
 * Get all temporal keywords for a specific year
 */
export function getKeywordsByYear(year: number): TemporalKeyword[] {
  switch (year) {
    case 2009: return KEYWORDS_2009;
    case 2010: return KEYWORDS_2010;
    case 2011: return KEYWORDS_2011;
    case 2012: return KEYWORDS_2012;
    case 2013: return KEYWORDS_2013;
    default: return [];
  }
}

/**
 * Get temporal keywords for a year range
 */
export function getKeywordsByYearRange(startYear: number, endYear: number): TemporalKeyword[] {
  const keywords: TemporalKeyword[] = [];
  for (let year = startYear; year <= endYear; year++) {
    keywords.push(...getKeywordsByYear(year));
  }
  return keywords;
}

/**
 * Get keywords by category across all years
 */
export function getKeywordsByCategory(category: TemporalKeyword['category']): TemporalKeyword[] {
  const allKeywords = [
    ...KEYWORDS_2009,
    ...KEYWORDS_2010,
    ...KEYWORDS_2011,
    ...KEYWORDS_2012,
    ...KEYWORDS_2013,
  ];
  
  return allKeywords.filter(k => k.category === category);
}

/**
 * Get high-relevance keywords (relevance >= threshold)
 */
export function getHighRelevanceKeywords(threshold: number = 0.7): TemporalKeyword[] {
  const allKeywords = [
    ...KEYWORDS_2009,
    ...KEYWORDS_2010,
    ...KEYWORDS_2011,
    ...KEYWORDS_2012,
    ...KEYWORDS_2013,
  ];
  
  return allKeywords
    .filter(k => k.relevance >= threshold)
    .sort((a, b) => b.relevance - a.relevance);
}

/**
 * Generate passphrase combinations using temporal keywords
 */
export function generateTemporalCombinations(
  basePhrase: string,
  year?: number
): string[] {
  const combinations: string[] = [];
  const keywords = year ? getKeywordsByYear(year) : getHighRelevanceKeywords(0.6);
  
  for (const kw of keywords) {
    // Add keyword before base phrase
    combinations.push(`${kw.keyword} ${basePhrase}`);
    
    // Add keyword after base phrase
    combinations.push(`${basePhrase} ${kw.keyword}`);
    
    // Add year suffix
    combinations.push(`${basePhrase}${year || kw.year}`);
    
    // Combined with year
    combinations.push(`${kw.keyword}${year || kw.year}`);
  }
  
  return combinations.slice(0, 100); // Limit to top 100
}

/**
 * Get all temporal keywords sorted by relevance
 */
export function getAllKeywordsSorted(): TemporalKeyword[] {
  const allKeywords = [
    ...KEYWORDS_2009,
    ...KEYWORDS_2010,
    ...KEYWORDS_2011,
    ...KEYWORDS_2012,
    ...KEYWORDS_2013,
  ];
  
  return allKeywords.sort((a, b) => b.relevance - a.relevance);
}

File: qig-backend/olympus/temporal_keywords.py
"""
Temporal Keyword Expansion Module

Generates hypotheses based on trending keywords during Bitcoin's early years.
Words and phrases that were culturally relevant during wallet creation
are more likely to have been used as passphrases.

Port of server/temporal-keywords.ts for use in Hephaestus hypothesis generation.
"""

from typing import List, Dict, Optional
from dataclasses import dataclass


@dataclass
class TemporalKeyword:
    """Represents a temporally-relevant keyword"""
    keyword: str
    year: int
    category: str  # 'politics', 'technology', 'pop_culture', 'economics', 'social', 'crypto'
    relevance: float  # 0-1, how likely to be used as passphrase
    context: str  # Why this was significant


# 2009 Temporal Keywords
KEYWORDS_2009 = [
    # Politics
    TemporalKeyword('obama', 2009, 'politics', 0.7, 'Obama inauguration January 2009'),
    TemporalKeyword('inauguration', 2009, 'politics', 0.5, 'Historic inauguration'),
    TemporalKeyword('change we can believe in', 2009, 'politics', 0.6, 'Obama campaign slogan'),
    TemporalKeyword('yes we can', 2009, 'politics', 0.6, 'Obama slogan'),
    
    # Economics
    TemporalKeyword('bailout', 2009, 'economics', 0.8, 'Bank bailouts, inspired Bitcoin'),
    TemporalKeyword('financial crisis', 2009, 'economics', 0.7, '2008-2009 crisis'),
    TemporalKeyword('recession', 2009, 'economics', 0.6, 'Great Recession'),
    TemporalKeyword('too big to fail', 2009, 'economics', 0.7, 'Bank phrase'),
    TemporalKeyword('lehman brothers', 2009, 'economics', 0.6, 'Bankruptcy 2008'),
    
    # Technology
    TemporalKeyword('iphone 3gs', 2009, 'technology', 0.5, 'Released June 2009'),
    TemporalKeyword('windows 7', 2009, 'technology', 0.5, 'Released October 2009'),
    TemporalKeyword('twitter', 2009, 'technology', 0.5, 'Growing rapidly'),
    TemporalKeyword('netbook', 2009, 'technology', 0.4, 'Popular device type'),
    
    # Pop Culture
    TemporalKeyword('avatar', 2009, 'pop_culture', 0.6, 'Movie released December 2009'),
    TemporalKeyword('michael jackson', 2009, 'pop_culture', 0.6, 'Died June 2009'),
    TemporalKeyword('swine flu', 2009, 'social', 0.5, 'H1N1 pandemic'),
    TemporalKeyword('miracle on the hudson', 2009, 'social', 0.5, 'January 2009 plane landing'),
    
    # Crypto
    TemporalKeyword('satoshi', 2009, 'crypto', 0.9, 'Bitcoin creator'),
    TemporalKeyword('genesis block', 2009, 'crypto', 0.8, 'First Bitcoin block'),
    TemporalKeyword('chancellor brink', 2009, 'crypto', 0.8, 'Genesis block headline'),
]

# 2010 Temporal Keywords
KEYWORDS_2010 = [
    # Technology
    TemporalKeyword('ipad', 2010, 'technology', 0.6, 'Released April 2010'),
    TemporalKeyword('instagram', 2010, 'technology', 0.5, 'Launched October 2010'),
    TemporalKeyword('foursquare', 2010, 'technology', 0.4, 'Popular check-in app'),
    
    # Crypto
    TemporalKeyword('pizza day', 2010, 'crypto', 0.8, '10,000 BTC for pizza May 22'),
    TemporalKeyword('laszlo', 2010, 'crypto', 0.6, 'Pizza buyer'),
    TemporalKeyword('bitcoin faucet', 2010, 'crypto', 0.7, 'Gavin Andresen faucet'),
    TemporalKeyword('slush pool', 2010, 'crypto', 0.6, 'First mining pool'),
    TemporalKeyword('wikileaks', 2010, 'crypto', 0.7, 'Bitcoin donations controversy'),
    
    # Pop Culture
    TemporalKeyword('inception', 2010, 'pop_culture', 0.5, 'Popular movie'),
    TemporalKeyword('fifa world cup', 2010, 'pop_culture', 0.4, 'South Africa 2010'),
    TemporalKeyword('vuvuzela', 2010, 'pop_culture', 0.4, 'World Cup horn'),
    
    # Social
    TemporalKeyword('deepwater horizon', 2010, 'social', 0.5, 'Oil spill disaster'),
    TemporalKeyword('haiti earthquake', 2010, 'social', 0.4, 'January 2010 disaster'),
]

# 2011 Temporal Keywords
KEYWORDS_2011 = [
    # Crypto
    TemporalKeyword('silk road', 2011, 'crypto', 0.7, 'Darknet market launched'),
    TemporalKeyword('one dollar', 2011, 'crypto', 0.7, 'BTC reached $1'),
    TemporalKeyword('parity', 2011, 'crypto', 0.6, 'BTC = USD parity'),
    TemporalKeyword('mtgox', 2011, 'crypto', 0.8, 'Major exchange'),
    
    # Social
    TemporalKeyword('arab spring', 2011, 'social', 0.6, 'Middle East uprisings'),
    TemporalKeyword('occupy wall street', 2011, 'social', 0.7, 'Financial protest movement'),
    TemporalKeyword('we are the 99', 2011, 'social', 0.6, 'Occupy slogan'),
    TemporalKeyword('fukushima', 2011, 'social', 0.5, 'Nuclear disaster'),
    TemporalKeyword('bin laden', 2011, 'politics', 0.5, 'Death May 2011'),
    
    # Technology
    TemporalKeyword('iphone 4s', 2011, 'technology', 0.5, 'Released October 2011'),
    TemporalKeyword('siri', 2011, 'technology', 0.5, 'Voice assistant debut'),
    TemporalKeyword('google plus', 2011, 'technology', 0.4, 'Launched June 2011'),
    
    # Pop Culture
    TemporalKeyword('game of thrones', 2011, 'pop_culture', 0.5, 'TV series premiered'),
    TemporalKeyword('steve jobs', 2011, 'pop_culture', 0.6, 'Died October 2011'),
]

# 2012 Temporal Keywords
KEYWORDS_2012 = [
    # Crypto
    TemporalKeyword('first halving', 2012, 'crypto', 0.8, 'November 2012'),
    TemporalKeyword('halving day', 2012, 'crypto', 0.7, 'Block reward cut to 25'),
    TemporalKeyword('asic mining', 2012, 'crypto', 0.7, 'ASIC miners announced'),
    TemporalKeyword('butterfly labs', 2012, 'crypto', 0.6, 'ASIC manufacturer'),
    TemporalKeyword('coinbase', 2012, 'crypto', 0.7, 'Exchange founded'),
    TemporalKeyword('bitcoin foundation', 2012, 'crypto', 0.6, 'Founded September'),
    
    # Social/Politics
    TemporalKeyword('mayan calendar', 2012, 'social', 0.5, '2012 apocalypse myth'),
    TemporalKeyword('gangnam style', 2012, 'pop_culture', 0.6, 'Viral sensation'),
    TemporalKeyword('hurricane sandy', 2012, 'social', 0.4, 'October 2012 storm'),
    TemporalKeyword('london olympics', 2012, 'pop_culture', 0.4, 'Summer 2012'),
]

# 2013 Temporal Keywords
KEYWORDS_2013 = [
    # Crypto
    TemporalKeyword('one thousand', 2013, 'crypto', 0.8, 'BTC hit $1000'),
    TemporalKeyword('cyprus crisis', 2013, 'crypto', 0.7, 'Bank crisis, BTC interest spike'),
    TemporalKeyword('silk road raid', 2013, 'crypto', 0.6, 'FBI shutdown October'),
    TemporalKeyword('ross ulbricht', 2013, 'crypto', 0.5, 'Silk Road arrest'),
    TemporalKeyword('bip32', 2013, 'crypto', 0.6, 'HD wallet standard'),
    TemporalKeyword('bip39', 2013, 'crypto', 0.6, 'Mnemonic standard'),
    
    # Pop Culture
    TemporalKeyword('snowden', 2013, 'politics', 0.7, 'NSA whistleblower'),
    TemporalKeyword('prism', 2013, 'politics', 0.6, 'NSA surveillance program'),
    TemporalKeyword('breaking bad', 2013, 'pop_culture', 0.5, 'Final season'),
    TemporalKeyword('bitcoin accepted here', 2013, 'crypto', 0.6, 'Merchant adoption growing'),
]


def get_keywords_by_year(year: int) -> List[TemporalKeyword]:
    """Get all temporal keywords for a specific year"""
    if year == 2009:
        return KEYWORDS_2009
    elif year == 2010:
        return KEYWORDS_2010
    elif year == 2011:
        return KEYWORDS_2011
    elif year == 2012:
        return KEYWORDS_2012
    elif year == 2013:
        return KEYWORDS_2013
    else:
        return []


def get_keywords_by_year_range(start_year: int, end_year: int) -> List[TemporalKeyword]:
    """Get temporal keywords for a year range"""
    keywords = []
    for year in range(start_year, end_year + 1):
        keywords.extend(get_keywords_by_year(year))
    return keywords


def get_keywords_by_category(category: str) -> List[TemporalKeyword]:
    """Get keywords by category across all years"""
    all_keywords = (
        KEYWORDS_2009 +
        KEYWORDS_2010 +
        KEYWORDS_2011 +
        KEYWORDS_2012 +
        KEYWORDS_2013
    )
    return [k for k in all_keywords if k.category == category]


def get_high_relevance_keywords(threshold: float = 0.7) -> List[TemporalKeyword]:
    """Get high-relevance keywords (relevance >= threshold)"""
    all_keywords = (
        KEYWORDS_2009 +
        KEYWORDS_2010 +
        KEYWORDS_2011 +
        KEYWORDS_2012 +
        KEYWORDS_2013
    )
    
    keywords = [k for k in all_keywords if k.relevance >= threshold]
    keywords.sort(key=lambda k: k.relevance, reverse=True)
    return keywords


def generate_temporal_combinations(
    base_phrase: str,
    year: Optional[int] = None,
    max_combinations: int = 100
) -> List[str]:
    """
    Generate passphrase combinations using temporal keywords
    """
    combinations = []
    keywords = get_keywords_by_year(year) if year else get_high_relevance_keywords(0.6)
    
    for kw in keywords:
        # Add keyword before base phrase
        combinations.append(f"{kw.keyword} {base_phrase}")
        
        # Add keyword after base phrase
        combinations.append(f"{base_phrase} {kw.keyword}")
        
        # Add year suffix
        combinations.append(f"{base_phrase}{year or kw.year}")
        
        # Combined with year
        combinations.append(f"{kw.keyword}{year or kw.year}")
    
    return combinations[:max_combinations]


def get_all_keywords_sorted() -> List[TemporalKeyword]:
    """Get all temporal keywords sorted by relevance"""
    all_keywords = (
        KEYWORDS_2009 +
        KEYWORDS_2010 +
        KEYWORDS_2011 +
        KEYWORDS_2012 +
        KEYWORDS_2013
    )
    
    all_keywords.sort(key=lambda k: k.relevance, reverse=True)
    return all_keywords


def get_crypto_specific_keywords() -> List[TemporalKeyword]:
    """Get only crypto-specific keywords (highest priority)"""
    return get_keywords_by_category('crypto')
